Fix Web Inspector: WI.ScopeBar.Event.SelectionChanged is fired four times whenever selecting an item (242705)

the-chenergy opened the pull request:
#### 2fd8234498119dd0369e13d87c7d7512b6d785be
<pre>
Fix Web Inspector: WI.ScopeBar.Event.SelectionChanged is fired four times whenever selecting an item (242705)
<a href="https://bugs.webkit.org/show_bug.cgi?id=242705">https://bugs.webkit.org/show_bug.cgi?id=242705</a>
<a href="https://rdar.apple.com/96963052">rdar://96963052</a>

Reviewed by Devin Rousso.

The bug report refers to an issue that happens with the components
ScopeBar and MultipleScopeBarItem, which is used across multiple
places in the inspector&apos;s frontend. This commit fixes the issue for both
components, such that the event `ScopeBar.Event.SelectionChanged` is
always only fired once, no matter whether MultipleScopeBarItem is
involved (like in the Sources tab) or not (like in the Console tab).

However, the solution to solving the bug for either component is
independent and different:
  - The class MultipleScopeBarItem already does well to suppress
    multiple echos of the SelectionChanged event with the help of
    `_ignoreItemSelectedEvent` (see <a href="https://github.com/WebKit/WebKit/blob/6d51d579af47028306a003653150ade5d395f942/Source/WebInspectorUI/UserInterface/Views/MultipleScopeBarItem.js#L211-L212)">https://github.com/WebKit/WebKit/blob/6d51d579af47028306a003653150ade5d395f942/Source/WebInspectorUI/UserInterface/Views/MultipleScopeBarItem.js#L211-L212)</a>
    However, in the original code, ScopeBar did not make use of this
    benefit and still listens to SelectionChanged from the individual
    items. When MultipleScopeBarItem is used, this commit makes ScopeBar
    listen to MultipleScopeBarItem instead.
  - When MultipleScopeBarItem is not used, ScopeBar in the original code
    did not do anything to protect against echos (i.e. events caused by
    the ScopeBar itself as a result of deselecting its ScopeBarItems).
    This commit mimics the fix from MultipleScopeBarItem and introduces
    the same boolean flag to bypass echos caused by the ScopeBar itself.

* Source/WebInspectorUI/UserInterface/Views/ScopeBar.js:
(WI.ScopeBar.prototype._populate):
  - Listen to SelectionChanged events from MultipleScopeBarItem instead
    of the individual ScopeBarItems.
(WI.ScopeBar.prototype._itemSelectionDidChange):
  - Guard against echos by ignoring events caused by
    `item.selected = ...` statements within the handler itself.
  - Rewrite the deselecting logic to be more readable.

Canonical link: <a href="https://commits.webkit.org/275435@main">https://commits.webkit.org/275435@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ffb5de0f1806cde5569fa71c42257efafa290313

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41740 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20754 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44122 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44316 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37833 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23891 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18085 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34497 "Found 1 new test failure: http/tests/scroll-to-text-fragment/no-scroll-after-stylesheet-load.html (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42314 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17680 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35938 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15167 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15372 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36960 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45719 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37924 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37296 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41052 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16555 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13591 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39492 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18174 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9374 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18231 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17818 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/11dca6d4eba7dc68d923fd5aa843e27a584077e7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24857

rcaliman-apple commented:
LGTM 👍

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ScopeBar.js:
```diff
@@ -146,8 +145,18 @@ WI.ScopeBar = class ScopeBar extends WI.NavigationItem
 
+    _multipleItemSelectionDidChange(event)
+    {
+        this.dispatchEventToListeners(WI.ScopeBar.Event.SelectionChanged);
```
this will likely also need to clear the selection from all other `WI.ScopeBarItem`

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c0cef0ab66ff4f72a87e482dc079a22d43131b5a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24857

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0baa1be486c317e2064d2fc955e8f20711e7e60e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24857

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ScopeBar.js:
```diff
@@ -148,20 +154,23 @@ WI.ScopeBar = class ScopeBar extends WI.NavigationItem
-            for (var i = 0; i < this._items.length; ++i) {
-                item = this._items[i];
+            // An exclusive item was selected, deselect everything else.
+            for (var item of this._ungroupedItems) {
```
rather than keep a second list in `_ungroupedItems` i think we could just do this
```js
for (let item of this._items) {
    if (this._shouldGroupNonExclusiveItems && !item.exclusive)
        continue;

    if (item !== sender)
        item.selected = false;
}

if (this._multipleItem && this._multipleItem !== sender)
    this._multipleItem.selected = false;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ScopeBar.js:
```diff
@@ -148,20 +154,23 @@ WI.ScopeBar = class ScopeBar extends WI.NavigationItem
-            for (var i = 0; i < this._items.length; ++i) {
-                item = this._items[i];
+            let replacesCurrentSelection = !event.data.extendSelection;
+            for (var item of this._ungroupedItems) {
```
ditto

dcrousso reviewed (approved):
nice fix!  only one minor suggestion :)

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/8ec5fc15b2e7b54823e977e88b7b7abdf5682c15)<details>

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41560 "Passed style check") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20574 "Build is in progress. Recent messages:OS: Sonoma (14.0), Xcode: 15.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43938 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44129 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/WinCairo-EWS "Waiting in queue, processing has not started yet") 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23683 "Build is in progress. Recent messages:OS: Sonoma (14.0), Xcode: 15.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17904 "Build is in progress. Recent messages:OS: Sonoma (14.1), Xcode: 15.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/44129 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42134 "Build is in progress. Recent messages:OS: Unknown (10.15.7), Xcode: 11.7") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/23683 "Build is in progress. Recent messages:OS: Sonoma (14.0), Xcode: 15.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/14/builds/43938 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/44129 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/23683 "Build is in progress. Recent messages:OS: Sonoma (14.0), Xcode: 15.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/14/builds/43938 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45500 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/23683 "Build is in progress. Recent messages:OS: Sonoma (14.0), Xcode: 15.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/14/builds/43938 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/45500 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request") | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/51/builds/17904 "Build is in progress. Recent messages:OS: Sonoma (14.1), Xcode: 15.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/45500 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request") | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/14/builds/43938 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-10-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-10-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | </details>

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ScopeBar.js:
```diff
@@ -148,26 +150,28 @@ WI.ScopeBar = class ScopeBar extends WI.NavigationItem
+
+        this._ignoreItemSelectedEvent = true;
+
+        const deselectOtherItems = () => {
```
i dont think we need to make this into a function since there's only one `return`, so we can simplify it as such
```js
if (event.target.selected) {
    let extendSelection = (event.data.extendSelection && !event.target.exclusive) || event.target === this._multipleItem;

    for (let item of this._items) {
        if (item === event.target)
            continue;

        if (extendSelection && !item.exclusive)
            continue;

        item.selected = false;
    }
}
```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ffb5de0f1806cde5569fa71c42257efafa290313. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24857

the-chenergy committed:
Fix Web Inspector: WI.ScopeBar.Event.SelectionChanged is fired four times whenever selecting an item (242705)
https://bugs.webkit.org/show_bug.cgi?id=242705
rdar://96963052

Reviewed by Devin Rousso.

The bug report refers to an issue that happens with the components
ScopeBar and MultipleScopeBarItem, which is used across multiple
places in the inspector's frontend. This commit fixes the issue for both
components, such that the event `ScopeBar.Event.SelectionChanged` is
always only fired once, no matter whether MultipleScopeBarItem is
involved (like in the Sources tab) or not (like in the Console tab).

However, the solution to solving the bug for either component is
independent and different:
  - The class MultipleScopeBarItem already does well to suppress
    multiple echos of the SelectionChanged event with the help of
    `_ignoreItemSelectedEvent` (see https://github.com/WebKit/WebKit/blob/6d51d579af47028306a003653150ade5d395f942/Source/WebInspectorUI/UserInterface/Views/MultipleScopeBarItem.js#L211-L212)
    However, in the original code, ScopeBar did not make use of this
    benefit and still listens to SelectionChanged from the individual
    items. When MultipleScopeBarItem is used, this commit makes ScopeBar
    listen to MultipleScopeBarItem instead.
  - When MultipleScopeBarItem is not used, ScopeBar in the original code
    did not do anything to protect against echos (i.e. events caused by
    the ScopeBar itself as a result of deselecting its ScopeBarItems).
    This commit mimics the fix from MultipleScopeBarItem and introduces
    the same boolean flag to bypass echos caused by the ScopeBar itself.

* Source/WebInspectorUI/UserInterface/Views/ScopeBar.js:
(WI.ScopeBar.prototype._populate):
  - Listen to SelectionChanged events from MultipleScopeBarItem instead
    of the individual ScopeBarItems.
(WI.ScopeBar.prototype._itemSelectionDidChange):
  - Guard against echos by ignoring events caused by
    `item.selected = ...` statements within the handler itself.
  - Rewrite the deselecting logic to be more readable.

Canonical link: https://commits.webkit.org/275435@main

webkit-commit-queue commented:
Committed 275435@main (2fd823449811): <https://commits.webkit.org/275435@main>

Reviewed commits have been landed. Closing PR #24857 and removing active labels.
